### DISCUSSION TITLE
How can I stop the http client from urlencoding double quotes? In Symfony 6.

### DISCUSSION THREAD

**Discussion by jerrac** [symfony/symfony#52525] [Q&A] [answered]
I'm trying to work with an API that apparently doesn't decode double quotes the way it should.

Specifically the `GET /admin/realms/{realm}/groups` endpoint documented here: https://www.keycloak.org/docs-api/22.0.5/rest-api/index.html#_groups 

As far as I can tell, I need to put the double quotes around the search parameter's value in order for it to do an exact match instead of a starts with (I think) match.

Unfortunately Symfony's http client likes to turn my urls from `http://keycloak:8080/admin/realms/master/groups?search="testing"` into`http://keycloak:8080/admin/realms/master/groups?search=%22testing%22"`, and Keycloak just returns no results.

While searching for an answer I think I found that older versions of Symfony let you turn off encoding, but newer ones don't? Is that true?

Anyone have an idea for a work around?

Thanks in advance.

**Comment by ghost** [comment]
You can verify the HttpClient source code and see if the line/function that does URL encoding uses the logic that is configurable or it is hardcoded.

**Reply by jerrac** [reply]
So, there are a lot of http clients, and it isn't clear which one the http_client service uses by default. But I think they all use this method: https://github.com/symfony/http-client/blob/0314e2d49939a9831929d6fc81c01c6df137fd0a/HttpClientTrait.php#L709C15-L709C15 

That means that there is no way to stop symfony from urlencoding my quotes. Assuming I followed the correct code path in PHPStorm... 

So that part of my post is cleared, but I could still use advice on how to get around the issue overall. Ideally without having to rewrite everything to not use symfony's http-client....

**Reply by ghost** [reply]
Maybe the author of http-client could help here, i.e. @nicolas-grekas

**Comment by nicolas-grekas** [comment]
I suppose you are using the `query` option? If yes, try using concatenation to build your final URL instead.

**Reply by jerrac** [reply]
Concatenating the url string is what I am doing.

```php
// ...
$queryString = '';
foreach ($queryParameters as $key => $value) {
    $useKey = urlencode($key);
    $useValue = urlencode($value);
    $queryString .= '&' . $useKey . '="' . $useValue . '"';
}
$query = ltrim($queryString, '&');
$url .= '?' . $query;
// ...
$request = $this->client->request($httpMethod, $url, $options);
// ...
```
I specifically don't run the whole string through urlencode, just the parts that might not be safe sometime in the future.... But if I'm reading the HttpClientTrait code correctly, the url gets parsed out into pieces, and then those pieces get run through urlencode before being put back together. 

Which, all things considered, is the way to do it. Ideally the solution would be fore the API to actually support urlencoded quotes properly, but that'd take a while to get done even if the maintainers agreed to do it. Hence my search for a work around.

**Reply by nicolas-grekas** [reply, marked as answer]
So, the encoding happens [here](https://github.com/symfony/http-client/blob/0314e2d49939a9831929d6fc81c01c6df137fd0a/HttpClientTrait.php#L662).

But double quotes are not allowed in URLs so you'd better report this on keycloak first and see how this goes there.

**Reply by jerrac** [reply]
Ok, so, I think I was mistaken in how you are supposed to get exact matches. When I started looking up the issue I had found the double quote idea in so I could reference it in my post, I found a pull request where they added an "exact" option you can pass in to get an exact match. That solved my problem. 

Thanks for taking the time to talk this out a bit. It helped!
